src/eos-prepare-usb-update.c

```diff
@@ -0,0 +1,665 @@
+                   raw_path,
+                   refspec->ref);
+      return FALSE;
+    }
+
+  if (!ostree_repo_resolve_rev (repo,
+                                refspec->str,
+                                FALSE,
+                                &ref_commit_id,
+                                error))
+    return FALSE;
+
+  while (g_strcmp0 (commit_id, ref_commit_id) != 0)
+    {
+      g_autoptr(GVariant) ref_commit = NULL;
+
+      if (!ostree_repo_load_commit (repo,
+                                    ref_commit_id,
+                                    &ref_commit,
+                                    NULL,
+                                    error))
+        return FALSE;
+      g_free (ref_commit_id);
+      ref_commit_id = ostree_commit_get_parent (ref_commit);
+      if (ref_commit_id == NULL)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                       "Commit %s is not reachable from refspec %s",
+                       commit_id,
+                       refspec->str);
+          return FALSE;
+        }
+    }
+
+  return TRUE;
+}
+
+static gboolean
+create_usb_repo (OstreeRepo *repo,
+                 EosRefspec *refspec,
+                 GFile *usb_path,
+                 GCancellable *cancellable,
+                 OstreeRepo **out_usb_repo,
+                 GError **error)
+{
+  g_autoptr(GFile) usb_repo_path = NULL;
+  g_autoptr(OstreeRepo) usb_repo = NULL;
+  g_autofree gchar *url = NULL;
+  g_auto(GVariantBuilder) builder = { { { 0, } } };
+  g_autoptr(GVariant) options = NULL;
+  g_autofree gchar *trusted_keys_name = NULL;
+  g_autoptr(GFile) remote_trusted_keys = NULL;
+  g_autoptr(GFileInputStream) gpg_stream = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  usb_repo_path = g_file_get_child (usb_path, "eos-update");
+  usb_repo = ostree_repo_new (usb_repo_path);
+  if (!ostree_repo_create (usb_repo,
+                           OSTREE_REPO_MODE_ARCHIVE_Z2,
+                           cancellable,
+                           error))
+    return FALSE;
+
+  if (!ostree_repo_remote_get_url (repo,
+                                   refspec->remote,
+                                   &url,
+                                   error))
+    return FALSE;
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE_VARDICT);
+  g_variant_builder_add (&builder, "{s@v}",
+                         "branches",
+                         g_variant_new_variant (g_variant_new_strv (&refspec->ref, 1)));
+  options = g_variant_ref_sink (g_variant_builder_end (&builder));
+  // ostree remote add --repo=dir/eos-update --no-gpg-verify eos https://endless:upgtnFSxLoDfUJnt@ostree.endlessm.com/ostree/eosfree-i386 eos2/i386
+  if (!ostree_repo_remote_add (usb_repo,
+                               refspec->remote,
+                               url,
+                               options,
+                               cancellable,
+                               error))
+    return FALSE;
```

Comment:
Adding the remote to the USB repo seems a little superfluous, but I guess this gives you the opportunity to later update the USB repository independently.